packages/mongodb-ts-autocomplete/src/index.ts

```diff
@@ -0,0 +1,215 @@
+      }
+    );
+
+    return `{
+  ${collectionProperties.join('\n')}
+}`;
+  }
+}
+
+class ConnectionSchema {
+  private readonly databaseNames: Set<string>;
+  private readonly databaseSchemas: Record<string, DatabaseSchema>;
+
+  constructor() {
+    // TODO: this is kinda the only real reason for this class: So we can keep
+    // track of the known databaseNames for a connection. It could be all the
+    // names from a listDatabases() or it could just be all the ones we've
+    // auto-completed for. The schemas can come from the autocompletion context.
+    // This can probably be added to the autocompletion context.
+    this.databaseNames = new Set();
+    this.databaseSchemas = Object.create(null);
+  }
+
+  addDatabase(databaseName: string) {
+    this.databaseNames.add(databaseName);
+    if (!this.databaseSchemas[databaseName]) {
+      this.databaseSchemas[databaseName] = new DatabaseSchema();
+    }
+  }
+
+  setDatabaseCollectionNames(databaseName: string, collectionNames: string[]) {
+    this.addDatabase(databaseName);
+    this.databaseSchemas[databaseName].setCollectionNames(collectionNames);
+  }
+
+  addCollectionSchema(
+    databaseName: string,
+    collectionName: string,
+    collectionSchema: JSONSchema
+  ) {
+    this.addDatabase(databaseName);
+    this.databaseSchemas[databaseName].setCollectionSchema(
+      collectionName,
+      collectionSchema
+    );
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const databaseProperties = [...this.databaseNames.values()].map(
+      (databaseName) => {
+        const def = this.databaseSchemas[databaseName]
+          ? this.databaseSchemas[databaseName].toTypescriptTypeDefinition()
+          : `{}`;
+        return `      '${databaseName}': ShellAPI.Database & ${def}`;
+      }
+    );
+
+    return `{
+  ${databaseProperties.join('\n')}
+}`;
+  }
+}
+
+type AutocompleteOptions = {
+  connectionId: string;
+  databaseName: string;
+  position?: number;
+};
+
+export default class MongoDBAutocompleter {
+  private readonly context: AutocompletionContext;
+  private connectionSchemas: Record<string, ConnectionSchema>;
+  private readonly autocompleter: Autocompleter;
+
+  constructor({ context, autocompleterOptions }: MongoDBAutocompleterOptions) {
+    this.context = context;
+    this.autocompleter = new Autocompleter(autocompleterOptions);
+
+    this.connectionSchemas = Object.create(null);
+    this.connectionSchemas;
+
+    this.autocompleter.updateCode(autocompleteTypes);
+  }
+
+  addConnection(connectionId: string): ConnectionSchema {
+    if (!this.connectionSchemas[connectionId]) {
+      this.connectionSchemas[connectionId] = new ConnectionSchema();
+    }
+    return this.connectionSchemas[connectionId];
+  }
+
+  getConnectionCode(connectionKey: string): string {
+    return `
+import * as ShellAPI from '/shell-api.ts';
+import * as bson from '/bson.ts';
+
+export type ServerSchema = ${this.connectionSchemas[
```

Comment:
Do we know if the compiler behaves differently depending on whether we dump all of that in a huge file vs generate multiple small files instead? We could experiment with generating a file per collection and then importing all of those in the connection file. That way, if we need to add a new collection, we'd only need to regenerate a subset of the entire code.

Reply:
The problem then is that we have to figure out how to convert from valid mongodb collection names to valid typescript type names. The benefit of having a giant per-server schema is that we can address it with `['database or collection name']`.

I started having one per collection and per database and per connection, then simplified it like this when the idea came from a thread in slack: https://mongodb.slack.com/archives/C08EXTB42LB/p1742298506300629

But I'm not convinced about this to the point that I can't be persuaded either way :)